state/accountsDB.go

```diff
@@ -1102,56 +1102,58 @@ func (adb *AccountsDB) SnapshotState(rootHash []byte) {
 
 	trieStorageManager.EnterPruningBufferingMode()
 
+	errChan := make(chan error, 1)
 	stats := newSnapshotStatistics(1)
 	go func() {
 		leavesChannel := make(chan core.KeyValueHolder, leavesChannelSize)
 		stats.NewSnapshotStarted()
-		trieStorageManager.TakeSnapshot(rootHash, rootHash, leavesChannel, stats, epoch)
-		adb.snapshotUserAccountDataTrie(true, rootHash, leavesChannel, stats, epoch)
+		trieStorageManager.TakeSnapshot(rootHash, rootHash, leavesChannel, errChan, stats, epoch)
+		adb.snapshotUserAccountDataTrie(true, rootHash, leavesChannel, errChan, stats, epoch)
 		trieStorageManager.ExitPruningBufferingMode()
 
 		stats.wg.Done()
 	}()
 
-	go adb.markActiveDBAfterSnapshot(stats, rootHash, "snapshotState user trie", epoch)
+	go adb.markActiveDBAfterSnapshot(stats, errChan, rootHash, "snapshotState user trie", epoch)
 
 	adb.waitForCompletionIfRunningInImportDB(stats)
 }
 
-func (adb *AccountsDB) markActiveDBAfterSnapshot(stats *snapshotStatistics, rootHash []byte, message string, epoch uint32) {
-	printStats(stats, message, rootHash)
+func (adb *AccountsDB) markActiveDBAfterSnapshot(stats *snapshotStatistics, errChan chan error, rootHash []byte, message string, epoch uint32) {
+	stats.PrintStats(message, rootHash)
 
 	trieStorageManager := adb.mainTrie.GetStorageManager()
-	if trieStorageManager.IsClosed() {
-		log.Debug("will not set activeDB in epoch as the snapshot might be incomplete since trie storage manager is closed", "epoch", epoch)
+	containsErrorDuringSnapshot := emptyErrChanReturningHadContained(errChan)
+	shouldNotMarkActive := trieStorageManager.IsClosed() || containsErrorDuringSnapshot
+	if shouldNotMarkActive {
+		log.Debug("will not set activeDB in epoch as the snapshot might be incomplete",
+			"epoch", epoch, "trie storage manager closed", trieStorageManager.IsClosed(),
+			"errors during snapshot found", containsErrorDuringSnapshot)
 		return
 	}
 
 	log.Debug("set activeDB in epoch", "epoch", epoch)
-	errPut := trieStorageManager.PutInEpoch([]byte(common.ActiveDBKey), []byte(common.ActiveDBVal), epoch)
+	errPut := trieStorageManager.PutInEpochWithoutCache([]byte(common.ActiveDBKey), []byte(common.ActiveDBVal), epoch)
 	handleLoggingWhenError("error while putting active DB value into main storer", errPut)
 }
 
-func printStats(stats *snapshotStatistics, identifier string, rootHash []byte) {
-	stats.wg.Wait()
-
-	stats.mutex.RLock()
-	defer stats.mutex.RUnlock()
-
-	log.Debug("snapshot statistics",
-		"type", identifier,
-		"duration", time.Since(stats.startTime).Truncate(time.Second),
-		"total num of nodes", stats.numNodes,
-		"total size", core.ConvertBytes(stats.trieSize),
-		"num data tries", stats.numDataTries,
-		"rootHash", rootHash,
-	)
+func emptyErrChanReturningHadContained(errChan chan error) bool {
+	contained := false
+	for {
+		select {
+		case <-errChan:
+			contained = true
```

Comment:
if we log them here we will end up logging twice. I have not changed the existing behavior since the logging is done in other parts and might contain more context information. I would leave this as it is now.
emptyErrChanReturningHadContained specify what it does: emptying the channel and returning true if the chan had contained an element. Do you have a better naming for this?

Reply:
I really can't come up with smth "decent" (errChanHadErrorWhileEmptying? -> sounds wrong). Could leave it as it is